wurst/math/Matrices.wurst

```diff
@@ -0,0 +1,352 @@
+package Matrices
+import NoWurst
+import public Vectors
+import _Primitives
+import Wurstunit
+
+/* Great thanks to euclideanspace.com for wide explanations. */
+
+
+/** Matrix 2x2
+00 01
+10 11
+*/
+public tuple mat2(real m00, real m01, real m10, real m11)
+
+public constant ZERO22 = mat2(0, 0, 0, 0)
+
+public constant IDENTITY22 = mat2(
+    1, 0,
+    0, 1)
+
+tuple inverseresult22(bool success, mat2 matrix)
+
+/** Matrix 3x3
+00 01 02
+10 11 12
+20 21 22
+*/
+public tuple mat3(  real m00, real m01, real m02,
+                    real m10, real m11, real m12,
+                    real m20, real m21, real m22)
+
+public constant ZERO33 = mat3(0, 0, 0, 0, 0, 0, 0, 0, 0)
+
+public constant IDENTITY33 = mat3(
+    1, 0, 0,
+    0, 1, 0,
+    0, 0, 1)
+
+tuple inverseresult33(bool success, mat3 matrix)
+
+/* =========== 2X2 MATRIX =========== */
+             
+public function mat2.op_plus(mat2 m) returns mat2
+    return mat2(this.m00 + m.m00,   this.m01 + m.m01,
+                this.m10 + m.m10,   this.m11 + m.m11)
+
+public function mat2.op_minus(mat2 m) returns mat2
+    return mat2(this.m00 - m.m00,   this.m01 - m.m01,
+                this.m10 - m.m10,   this.m11 - m.m11)
+
+public function mat2.op_minus(real scalar) returns mat2
+    return mat2(this.m00 - scalar,  this.m01 - scalar,
+                this.m10 - scalar,  this.m11 - scalar)
+
+public function mat2.op_plus(real scalar) returns mat2
+    return mat2(this.m00 + scalar,  this.m01 + scalar,
+                this.m10 + scalar,  this.m11 + scalar)
+    
+public function mat2.op_mult(real scalar) returns mat2
+    return mat2(this.m00 * scalar,    this.m01 * scalar,
+                this.m10 * scalar,    this.m11 * scalar)
+
+public function real.op_mult(mat2 m) returns mat2
+    return m * this
+
+/** Matrix multiplication is not commutative that means matrix*vect is not the same as vec*matrix. */
+public function mat2.op_mult(vec2 v) returns vec2
+    return vec2(this.m00 * v.x + this.m01 * v.y, this.m10 * v.x + this.m11 * v.y)
+
+/** Matrix multiplication is not commutative that means matrix*vect is not the same as vec*matrix. */
+public function vec2.op_mult(mat2 m) returns vec2
+    return vec2(this.x * m.m00 + this.y * m.m10, this.x * m.m01 + this.y * m.m11)
+
+public function mat2.op_mult(mat2 m) returns mat2
+    return mat2(this.m00 * m.m00 + this.m01 * m.m10,    this.m00 * m.m01 + this.m01 * m.m11, 
+                this.m10 * m.m00 + this.m11 * m.m10,    this.m10 * m.m01 + this.m11 * m.m11)
+
+/** Returns matrix' column by id as a vector or zero-vector if doesn't exist. */
+public function mat2.col(int id) returns vec2
+    vec2 result
+    switch id
+        case 0
+            result = vec2(this.m00, this.m10)
+        case 1
+            result = vec2(this.m01, this.m11)
+        default
+            result = ZERO2

```

Comment:
IMO, this is an invalid use of the API and should result in an error instead of a zero-value being returned.

Reply:
Such functions (as .inverse ones) should return `null` if result can't be obtained. Since we use tuples which can't take `null` and since WC3 doesn't support exceptions how can we do it here? `inversereulst` was a good solution for `.inverse` functions but such approach would be surplus here.
I completely agree with you, but I have no idea.

Reply:
Just call `error()`.

I don't see any use-case where using this function with `n` out of bounds would result in any kind of 'desirable' result.

When calculating an inverse, you don't know ahead of time if there's one or not, so we return a result tuple.

When getting a row, however, you *do* know if there's a row or not, so passing in an invalid value is a mis-use of the API.

Hence, just call `error()` specifying the error.

I might submit a PR some time later with some kind of exception-handling mechanism for use instead of error(), since all the infrastructure is there already, but for now, just use `error()`.

Reply:
> When calculating an inverse, you don't know ahead of time if there's one or not, so we return a result tuple.

We can find it from pre-calculation of the determinant and then compare obtained result with 0. But it gives us many extra calculations. Moreover, since the role of this package, obviously, is rotations we always know there's the inverse matrix and we can obtain it through `.transpose` without any calculations.

Meh, just lyric.

I agree and will add to the code `error` calls. Obviously is much better to know where the error from.